### ISSUE TITLE
Found AssertionError

### ISSUE THREAD

**Issue by jyhflash** [kangshwan/MotionMamba_Implementation#1]
Hi, thanks for your code, I run the code but I found
assert residual.shape == x_shape_og
AssertionError

At the end of Encoder blocker,then next to the mixer(transformer blocker), I found in the motion_mamba1_denoiser.py

hidden_state = torch.cat((sample, emb_latent), axis=1)
out = self.mixer(query = hidden_state, key = emb_latent, value = emb_latent)[0]
out = torch.cat((out, emb_latent), axis=1)

I found you cat twice before and after entering the transformer, is that true?Because this cause AssertionError.

**Comment by kangshwan** [comment]
Oops, my mistake about the code.

I'm running several tests on the mixer block, comparing results with and without slicing.

In my code file `motion_mamba1_denoiser.py`:

- **Comment out** line **1355**.
- **Uncomment** line **1357**, specifically the line:  
  ```python
  hidden_state = self.query_pos(hidden_state[:, :seqlen])

This change should work. If there are still issues, please let me know again.

**Comment by kangshwan** [comment]
I added emb to latent(sample) due to the lack of contextual information. Initially, I attempted mixing emb_latent directly within the mixer block, but the results were poor.

In the MLD code and paper, the authors mentioned that concatenating emb into latent yields better performance, so I adopted that approach for motion_mamba. Consequently, the dimension of the hidden_state has increased.

Regarding the mixer block, it seemed like emb_latent was being redundantly provided, so I split it by seqlen and fed it accordingly. This adjustment worked in my case, and I was able to achieve some promising results. I'll update the README file soon.

Currently, I'm testing an alternative approach where I avoid splitting the hidden_state. When slicing the hidden_state, I had to concatenate emb_latent again before passing it through the decoder, which I'm trying to simplify.

Thank you for your attention. Feel free to ask if you have any questions!

**Comment by jyhflash** [comment]
> Oops, my mistake about the code.
> 
> I'm running several tests on the mixer block, comparing results with and without slicing.
> 
> I think the error occurred because I committed the code in the middle of making changes.
> 
> In my code file `motion_mamba1_denoiser.py`:
> 
> * **Comment out** line **1355**.
> * **Uncomment** line **1357**, specifically the line:
>   ```python
>   hidden_state = self.query_pos(hidden_state[:, :seqlen])
>   ```
> 
> This change should work. If there are still issues, please let me know again.

oh, it works, thank you very much.

**Comment by jyhflash** [comment]
> **Additional Comments**
> 
> I added the `emb_latnet` to `latent(sample)` due to a lack of contextual information. Initially, I attempted to mix `emb_latent` directly within the mixer block, but the results were poor.
> 
> In the code and paper for MLD, the author mentioned that concatenating the `emb` with the latent representation performs well, so I adopted that approach in `motion_mamba`. As a result, the dimension of the `hidden_state` increased.
> 
> Regarding the mixer block, I felt that `emb_latent` was being redundantly provided, so I sliced it along the sequence length (`seqlen`) before passing it forward. This adjustment worked in my case, and I achieved some results. I will update the README file later with further details.
> 
> Currently, I am testing an approach where the `hidden_state` is not split. The reason is that when slicing the `hidden_state`, I have to concatenate `emb_latent` again before passing it through the decoder.
> 
> Thank you for your attention, and feel free to ask if you have any questions.

appreciate for your answer, do you think emb_latent is needed to input the mixer block?
I think about it, if is needed, I uncomment the line1365,
out = torch.cat((out, emb_latent), axis=1)
and it can also run.
and that the emb_latent can put in the mixer block, no need to slice the sequence.

**Comment by kangshwan** [comment]
I also think line 1365 might be unnecessary. I considered it from two perspectives:

1. **Aligning the residual shape (bs, 3, dim) from the Encoder by concatenating `emb_latent` to `out`**
   - This was done to maintain the residual connection. 
   - The reason for slicing with `seqlen` before entering the mixer block was based on the assumption that values had been sufficiently refined after passing through the Mamba block.
   - However, I believed that each residual output from the Encoder might retain more raw information, which I didn’t want to discard.
   - Therefore, to align the shapes, and retain raw information, I performed another concatenation at line 1365. Without line 1365, I remember lines 1369 and 1378 not functioning correctly. If that is not the case, please let me know.

2. **Performing residual connection slicing on the residual shape (bs, 3, dim) from the Encoder before passing it into the mixer block, similar to the previous slicing step**
   - I haven't tested this approach yet.
   - Personally, I think it should work since `emb_latent` was also unnecessary during the decoder process in MLD. However, my concern is the potential loss of meaningful information from the Encoder.

**Comment by jyhflash** [comment]
> I also think line 1365 might be unnecessary. I considered it from two perspectives:
> 
> 1. **Aligning the residual shape (bs, 3, dim) from the Encoder by concatenating `emb_latent` to `out`**
>    
>    * This was done to maintain the residual connection.
>    * The reason for slicing with `seqlen` before entering the mixer block was based on the assumption that values had been sufficiently refined after passing through the Mamba block.
>    * However, I believed that each residual output from the Encoder might retain more raw information, which I didn’t want to discard.
>    * Therefore, to align the shapes, and retain raw information, I performed another concatenation at line 1365. Without line 1365, I remember lines 1369 and 1378 not functioning correctly. If that is not the case, please let me know.
> 2. **Performing residual connection slicing on the residual shape (bs, 3, dim) from the Encoder before passing it into the mixer block, similar to the previous slicing step**
>    
>    * I haven't tested this approach yet.
>    * Personally, I think it should work since `emb_latent` was also unnecessary during the decoder process in MLD. However, my concern is the potential loss of meaningful information from the Encoder.
 I will consider your opinion,thank you!!

**Comment by jyhflash** [comment]
Hi, author, for the paper motion mamba, it mention the vae part, its loss only consit of MSE loss of latent noise, is it?

 
![屏幕截图 2025-01-13 093127](https://github.com/user-attachments/assets/c62bc220-c53d-440c-9e72-86e1043baea1)

**Comment by kangshwan** [comment]
Yes, my answer to your question is yes.

I implemented my code this way because that's how MLD trained the model. The original paper mentioned reusing the VAE from MLD, and in my code, the VAE training process is exactly the same as in MLD. MLD also trained the model using MSE.

**Comment by jyhflash** [comment]
> My answer to your question is yes.
> 
> I implemented my code this way because that's how MLD trained the model. The original paper mentioned reusing the VAE from MLD, and in my code, the VAE training process is exactly the same as in MLD. MLD also trained the model using MSE.

OK, Thank you. Can you show the accurate part of the loss mse of your code? cause I found in your code file /MotionMamba_Implementation-main/mld/models/losses/mld.py line 102
if stage in vae:
it have recons_feature, recons_joints and kl_motion loss?

**Comment by kangshwan** [comment]
Ah, sorry for my misunderstanding of your question.  

The paper mentioned using a VAE as proposed in MLD, so training the VAE follows the same procedure as described in the code.  

The line you highlighted refers to the loss function used for the Motion Mamba model.  

`Finally, our latent diffusion model is trained with an objective focusing on the minimization of MSE...`
The latent diffusion model acts as the Motion Mamba denoiser, and MSE is used to train this denoiser, not VAE.

**Comment by kangshwan** [comment]
This is the paper on MLD: [https://arxiv.org/pdf/2212.04048](https://arxiv.org/pdf/2212.04048).  
You can check how the VAE is structured in MLD paper appendix.  
Hope this helps!

**Comment by kangshwan** [comment]
![image](https://github.com/user-attachments/assets/b992fc09-ed99-421a-b300-59d1d74eb5c9)  
It's strange that the MLD mentions using both MSE and KL loss for training the VAE, but the code doesn't include any MSE loss.  
There have been no modifications to the VAE part—I adopted the entire process from MLD.  
I'm unsure whether they changed the method or not. 
As you mentioned, VAE typically uses MSE, but they seem to be using `SmoothL1Loss` instead.

**Comment by jyhflash** [comment]
> ![image](https://private-user-images.githubusercontent.com/46666296/402411532-b992fc09-ed99-421a-b300-59d1d74eb5c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3Mzc5NDcsIm5iZiI6MTczNjczNzY0NywicGF0aCI6Ii80NjY2NjI5Ni80MDI0MTE1MzItYjk5MmZjMDktZWQ5OS00MjFhLWIzMDAtNTlkMWQ3NGViNWM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEzVDAzMDcyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0MDRmNzA2ZGFkMDNhNTA5N2E5NjJkNGQwZGE0YTFhMzYwYzE4NzNkMmNhNTY5NTQ3YmEyNzFkN2EyMTRhNDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2eyMpvWRYbjV4YSWSBUo_0VCwDR2ngc91ss5eW_t7d0) It's strange that the MLD mentions using both MSE and KL loss for training the VAE, but the code doesn't include any MSE loss. There have been no modifications to the VAE part—I adopted the entire process from MLD. I'm unsure whether they changed the method or not. As you mentioned, VAE typically uses MSE, but they seem to be using `SmoothL1Loss` instead.

Ok, Thank you for your reply❤️! I think I will follow the code use SmoothL1Loss and KL loss.

**Comment by kangshwan** [comment]
If you want to follow the paper exactly, change line 82 from `torch.nn.SmoothL1Loss` to `nn.MSELoss`. 
However, I'm not sure if the code will work correctly or if the results will be as good as the current version.

**Comment by jyhflash** [comment]
> If you want to follow the paper exactly, change line 82 from `torch.nn.SmoothL1Loss` to `nn.MSELoss`. However, I'm not sure if the code will work correctly or if the results will be as good as the current version.

Ok, I will try it. I think the result may be not much difference.
